### ISSUE TITLE
GetFieldMappingsRequest() called from within an AbstractQueryBuilder returns an empty map

### ISSUE THREAD

**Issue by alexklibisz** [elastic/elasticsearch#64069]
**Elasticsearch version** (`bin/elasticsearch --version`): upgrading from 7.6.2 to 7.9.2

**Plugins installed**: [elastiknn](https://github.com/alexklibisz/elastiknn) (this issue is in the Elastiknn code)

**JVM version** (`java -version`): 14

**OS version** (`uname -a` if on a Unix-like system): Ubuntu

**Description of the problem including expected versus actual behavior**:

There's a part of the Elastiknn plugin that uses the client provided by the `QueryShardContext` (i.e. `context.getClient()`) to make a `GetFieldMappingsRequest()`. Specifically, a `new GetFieldMappingsRequest().indices(nameOfTheIndex).fields(nameOfTheField)`. The exact code is here: https://github.com/alexklibisz/elastiknn/blob/es-792/elastiknn-plugin/src/main/scala/com/klibisz/elastiknn/query/KnnQueryBuilder.scala#L145-L157. It's scala but should be pretty clear what's happening.

Prior to 7.9.2, this would return basically the same JSON-ish Java Map structure that you'd get from using the `GET _mapping?pretty` endpoint. Now it just returns a map with a single key for the index name and no further values. 

I've made sure that a curl request to the `GET _mapping?pretty` endpoint returns the correct mapping JSON.

The broader context here is that I'm updating the custom field mappers to account for changes made between 7.8.x and 7.9.x. Specifically, it looks like there were some changes to the MappedFieldType and FieldType constructs between those versions. Here's the exact diff: https://github.com/alexklibisz/elastiknn/pull/173/files#diff-0d7f259cbb9b71de0625d9bdb96c1fc7b4f896f2e33220b964279116320564b5

**Steps to reproduce**:

I can provide a repro but it will take some non-trivial effort to get all the right components in place, given it's a custom plugin.
I'm really hoping there's just some subtlety or tribal knowledge to how I make the request or how I define the custom mapper.
If not, I'll edit the issue with repro directions.

**Provide logs (if relevant)**:
N/A

**Comment by jtibshirani** [comment]
Hello @alexklibisz, we didn't make any intentional changes to the response format, and we certainly shouldn't be returning an empty response if there are field mappings.

Before jumping into a reproduction, I wanted to offer a high-level suggestion: we don't usually suggest making a client call while constructing a query. You should be able to look up relevant mapping information directly using `QueryShardContext`. For example it's common to use `QueryShardContext#getFieldType` to get a `MappedFieldType`. These `MappedFieldType` objects are designed to contain all the information needed for searching a field.

**Comment by alexklibisz** [comment]
Hi @jtibshirani , thanks for the quick response. I agree making a client call in that place was a bit hacky. I was actually only calling once and then caching the response for each index/field. Still hacky, so I'm open to revising it with a better approach. I mainly just need a way to access the original JSON used to construct the mapping (i.e. the same thing that's returned from `GET _mapping` API requests. I'll look into getFieldType later today.

**Comment by alexklibisz** [comment]
`QueryShardContext#getFieldType` doesn't seem to exist. I've tried using `QueryShardContext#fieldMapper`, and there is a way for me to inject the mapping definition to the `MappedFieldType` so that I can get it back out. But, my concern is I'm not sure what will happen if there are > 1 fields with that type, but with different parameters, in the same index. For example, I can have one `elastiknn_sparse_bool_vector` with parameter `{"dims": 100}` and another with parameter `{"dims": 999}`. Calling `fieldMapper("elastiknn_sparse_bool_vector")` returns one of them.. but I've no way to know which one it is.

**Comment by alexklibisz** [comment]
Update: I was able to retrieve the mapping using `QueryShardContext#fieldMapper`. I hadn't realized that the type name and the field name can be different. Once I started correctly setting the field name, I was able to retrieve the MappedFieldType for the field. I'm still not sure what the original issue was with the `GetFieldMappingsRequest()`. But just accessing the MappedFieldType is a simpler and better way to solve the problem for my case.
